instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/HostIdResourceProvider.java

```diff
@@ -0,0 +1,211 @@
+    this(
+        HostIdResourceProvider::getOsType,
+        HostIdResourceProvider::readMachineIdFile,
+        HostIdResourceProvider::queryWindowsRegistry);
+  }
+
+  // Visible for testing
+
+  HostIdResourceProvider(
+      Supplier<OsType> getOsType,
+      Function<Path, List<String>> machineIdReader,
+      Supplier<List<String>> queryWindowsRegistry) {
+    this.getOsType = getOsType;
+    this.machineIdReader = machineIdReader;
+    this.queryWindowsRegistry = queryWindowsRegistry;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+    OsType osType = getOsType.get();
+    switch (osType) {
+      case WINDOWS:
+        return readWindowsGuid();
+      case LINUX:
+        return readLinuxMachineId();
+      case OTHER:
+        break;
+    }
+    logger.fine("Unsupported OS type: " + osType);
+    return Resource.empty();
+  }
+
+  private Resource readLinuxMachineId() {
+    Path path = FileSystems.getDefault().getPath("/etc/machine-id");
+    List<String> lines = machineIdReader.apply(path);
+    if (lines.isEmpty()) {
+      return Resource.empty();
+    }
+    return Resource.create(Attributes.of(ResourceAttributes.HOST_ID, lines.get(0)));
+  }
+
+  private static List<String> readMachineIdFile(Path path) {
+    try {
+      List<String> lines = Files.readAllLines(path);
+      if (lines.isEmpty()) {
+        logger.fine("Failed to read /etc/machine-id: empty file");
+      }
+      return lines;
+    } catch (IOException e) {
+      logger.log(Level.FINE, "Failed to read /etc/machine-id", e);
+      return Collections.emptyList();
+    }
+  }
+
+  private static OsType getOsType() {
+    // see
+    // https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/SystemUtils.java
+    // for values
+    String osName = System.getProperty("os.name");
+    if (osName == null) {
+      return OsType.OTHER;
+    }
+    if (osName.startsWith("Windows")) {
+      return OsType.WINDOWS;
+    }
+    if (osName.toLowerCase(Locale.ROOT).equals("linux")) {
+      return OsType.LINUX;
+    }
+    return OsType.OTHER;
+  }
+
+  private Resource readWindowsGuid() {
+    List<String> lines = queryWindowsRegistry.get();
+
+    for (String line : lines) {
+      if (line.contains("MachineGuid")) {
+        String[] parts = line.trim().split("\\s+");
+        if (parts.length == 3) {
+          return Resource.create(Attributes.of(ResourceAttributes.HOST_ID, parts[2]));
+        }
+      }
+    }
+    logger.fine("Failed to read Windows registry: No MachineGuid found in output: " + lines);
+    return Resource.empty();
+  }
+
+  private static List<String> queryWindowsRegistry() {
+    try {
+      ProcessBuilder processBuilder = new ProcessBuilder("cmd", "/c", REGISTRY_QUERY);
+      processBuilder.redirectErrorStream(true);
+      Process process = processBuilder.start();
+
+      List<String> output = getProcessOutput(process);
+
+      if (!process.waitFor(2, TimeUnit.SECONDS)) {
```

Comment:
now that you are reading the output on the main thread you should remove the timeout and handling of it

Reply:
done